pkg/client/events_query/client.go

```diff
@@ -0,0 +1,273 @@
+// event messages matching the given query. It receives an either.Either[[]byte]
+// which is either an error or the event message bytes.
+// (see: https://pkg.go.dev/github.com/cometbft/cometbft/types#pkg-constants)
+// (see: https://docs.cosmos.network/v0.47/core/events#subscribing-to-events)
+func (eqc *eventsQueryClient) EventsBytes(
+	ctx context.Context,
+	query string,
+) (client.EventsBytesObservable, error) {
+	// Must (write) lock eventsBytesAndConnsMu so that we can safely check for
+	// existing subscriptions to the given query or add a new eventsBytes to the
+	// observableConns map.
+	// The lock must be held for both checking and adding to prevent concurrent
+	// calls to this function from racing.
+	eqc.eventsBytesAndConnsMu.Lock()
+	// Deferred (write) unlock.
+	defer eqc.eventsBytesAndConnsMu.Unlock()
+
+	// Check if an event subscription already exists for the given query.
+	if eventsBzConn := eqc.eventsBytesAndConns[query]; eventsBzConn != nil {
+		// If found it is returned.
+		return eventsBzConn.eventsBytes, nil
+	}
+
+	// Otherwise, create a new event subscription for the given query.
+	eventsBzConn, err := eqc.newEventsBytesAndConn(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	// Insert the new eventsBytes into the eventsBytesAndConns map.
+	eqc.eventsBytesAndConns[query] = eventsBzConn
+
+	// Unsubscribe from the eventsBytes when the context is done.
+	go eqc.goUnsubscribeOnDone(ctx, query)
+
+	// Return the new eventsBytes observable for the given query.
+	return eventsBzConn.eventsBytes, nil
+}
+
+// Close unsubscribes all observers from all event subscription observables.
+func (eqc *eventsQueryClient) Close() {
+	eqc.close()
+}
+
+// close unsubscribes all observers from all event subscription observables.
+func (eqc *eventsQueryClient) close() {
+	eqc.eventsBytesAndConnsMu.Lock()
+	defer eqc.eventsBytesAndConnsMu.Unlock()
+
+	for query, obsvblConn := range eqc.eventsBytesAndConns {
+		_ = obsvblConn.conn.Close()
+		obsvblConn.eventsBytes.UnsubscribeAll()
+
+		// remove closed eventsBytesAndConns
+		delete(eqc.eventsBytesAndConns, query)
+	}
+}
+
+// getNextRequestId increments and returns the JSON-RPC request ID which should
+// be used for the next request. These IDs are expected to be unique (per request).
+func (eqc *eventsQueryClient) getNextRequestId() string {
+	eqc.nextRequestId++
+	return fmt.Sprintf(requestIdFmt, eqc.nextRequestId)
+}
+
+// newEventwsBzAndConn creates a new eventsBytes and connection for the given query.
+func (eqc *eventsQueryClient) newEventsBytesAndConn(
+	ctx context.Context,
+	query string,
+) (*eventsBytesAndConn, error) {
+	conn, err := eqc.openEventsBytesAndConn(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct an eventsBytes for the given query.
+	eventsBzObservable, eventsBzPublishCh := channel.NewObservable[either.Either[[]byte]]()
+
+	// TODO_INVESTIGATE: does this require retry on error?
```

Comment:
Would the retry happen inside the function given that it's run as a goroutine?

Reply:
**TL;DR: I don't think we should do this here.**

I was comparing this to the block client implementation where we do a similar thing but are retrying on error. I deleted this TODO in the block client branch because I decided that we shouldn't retry here, I think it should be up to the observer how to handle an async error from the `#EventsBytes()` observable. 

Since the client unsubscribe's all observers (i.e. no more notifications) after encountering an error, observers should only ever see one error before they're unsubscribed. 

An example of this can currently be seen in the block client. In that case, receiving an async error from `#EventsBytes()` likely means that our long-running connection failed at some point, in which case we can simply restart it by calling `#EventsBytes()` again.

Reply:
#PUC with a link to this discussion.